### PR TITLE
Fix java date parsing to be compatible with joda

### PR DESCRIPTION
**Pull request by pgomulka**
ZonedDateTime toString format yyyy-MM-ddTHH:mmZ is failing parsing by
our java-time DateFormatters whereas it passes Joda parsing. The pattern
is strict_date_optional_time||epoch_millis

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 1]
```diff
@@ -85,13 +85,13 @@
         .optionalStart()
         .appendFraction(MILLI_OF_SECOND, 3, 3, true)
         .optionalEnd()
+        .optionalEnd()
```
Comment (pgomulka) [inline server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: this is just moving the zoneId below the optional section of seconds/milliseconds. Hard to see with this formatting

File: server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java [hunk 2]
```diff
@@ -390,6 +391,28 @@ public void testDuelingStrictParsing() {
 
         assertSameDate("2012-W31-5", "strict_weekyear_week_day");
         assertParseException("2012-W1-1", "strict_weekyear_week_day");
+//failing below
+        assertSameDate("2015-01-04T00:00Z", "strict_date_optional_time||epoch_millis");
+    }
+//Failing
```
Comment (pgomulka) [inline server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java]: will delete these if the PR is ok, but it helps investigating

File: server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java [hunk 3]
```diff
@@ -390,6 +391,28 @@ public void testDuelingStrictParsing() {
 
         assertSameDate("2012-W31-5", "strict_weekyear_week_day");
         assertParseException("2012-W1-1", "strict_weekyear_week_day");
+//failing below
+        assertSameDate("2015-01-04T00:00Z", "strict_date_optional_time||epoch_millis");
```
Comment (spinscale) [inline server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java]: can you just use `strict_date_optional_time` and remove `epoch_millis`? Also, can yo move this to the other `strict_date_optional_time` checks?

Can you also move this `strict_date_optional_time` checks to the others in this method?

File: server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java [hunk 4]
```diff
@@ -390,6 +391,28 @@ public void testDuelingStrictParsing() {
...
+    }
+//Failing
+    public void testFailingParseJavaTime(){
+        DateFormatter formatter = DateFormatters.forPattern("strict_date_optional_time||epoch_millis");
+        TemporalAccessor parse = formatter.parse("2015-01-04T00:00Z");
+    }
+    //Passing
+    public void testParsingJoda(){
```
Comment (spinscale) [inline server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java]: these two methods are not needed anymore, correct?

**Review by spinscale — CHANGES_REQUESTED** [review]

File: server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java (hunk 2 above)
Comment (spinscale) [reply server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java]: for sure I'm fine with the PR in general! :-)

File: server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java (hunk 4 above)
Comment (pgomulka) [reply server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java]: these are not needed. I only left them if you wanted to verify that quickly on your own

File: server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java (hunk 3 above)
Comment (pgomulka) [reply server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java]: absolutely - will clean this up

File: server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java [hunk 5]
```diff
@@ -33,6 +33,7 @@
 import java.util.Locale;
 
 import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.equalTo;
```
Comment (spinscale) [inline server/src/test/java/org/elasticsearch/common/joda/JavaJodaTimeDuellingTests.java]: unused import

**Review by spinscale — APPROVED** [review]
LGTM once CI is happy due to checkstyle. Thanks for finding this!

**Comment by pgomulka** [comment]
sorry about that @colings86 . totally slipped my mind. Thanks for fixing this

**Comment by colings86** [comment]
@pgomulka note that I didn't add the change type and version labels so could you please add those?

**Comment by spinscale** [comment]
@colings86 I do not think that this is necessary, as this PR is merged into a feature branch?

**Pull request merged**